Probe as thread

jan-cerny opened the pull request:
I have removed multiple processes.

I have converted the main() function of probe process that we have now to a probe thread that runs inside the oscap process.

For communication with probe, I have made use of fact that threads can share variables between them. This shared variable is a pair of queues that can hold data.
The AF_UNIX socket that was used to communicate between process is there replaced by 2 queues - one queue for messages from oscap to probe and another one to talk back.

mpreisler commented on src/OVAL/probes/probe/main.h:
```diff
@@ -20,13 +20,13 @@
-#ifndef MAIN_H
-#define MAIN_H
+#ifndef OPENSCAP_MAIN_H
+#define OPENSCAP_MAIN_H
```
`main.h` is asking for trouble, rename the file to probe_main.h please

jan-cerny commented:
@mpreisler Problem with this is that it fails tests/API/OVAL/report_variable_values tests, because some instances of texfilecontent items are missing in the output produced by this. I will try to figure out why is this happening. It might be possible that the probe thread terminates too soon in this prototype. I am going to investigate.

jan-cerny commented:
It looks like the probe reuses the data from past runs even if it shouldn't. I think it can be caused by our caches -- ncache is a global variable.

mpreisler commented:
@jan-cerny needs a rebase

jan-cerny commented:
Yes, I agree, I am going to rebase.

jan-cerny commented on src/OVAL/CMakeLists.txt:
```diff
@@ -64,6 +64,10 @@ if (ENABLE_PROBES)
 	"oval_probe_meta.h"
 	"probes/probe-api.c"
 	"probes/_probe-api.h"
+	"probes/fsdev.c"
```
This is a badly resolved merge. Those files are already moved to lines 74-78, they shouldn't be duplicated here.

jan-cerny commented on src/OVAL/probes/SEAP/sch_queue.c:
```diff
@@ -0,0 +1,191 @@
+{
+	int ret = 0;
+	sch_queuedata_t *data = (sch_queuedata_t *) desc->scheme_data;
+	if (pthread_cancel(data->probe_thread_id) != 0) {
```
I think using pthread_cancel is a wrong approach. It might be a pthread_join instead.

jan-cerny commented:
So far this prototype looks very promising. It already works for most of the use cases, it can handle OVAL variables and sets, it passes most of the tests.

However there are 2 major problems that aren't resolved:

1. It needs to build also if not all dependencies are installed on the system where it builds - when some libraries or header files are not available. This is main reason of Jenkins fail. Right now the prototype expects every dependency is available, because of how a list of probes in probe-table.c https://github.com/jan-cerny/openscap/blob/e042fca9b970153e96bb3f9a311da4a4a63b1c63/src/OVAL/probes/probe-table.c is done. I think that this is solvable. I think that CMake could generate some variables that we would then check as `#ifdefs` in probe-table.c. We would then try to use only the probes which have all the dependencies in place. But I am afraid that at the end of all days I will have to create a dedicated `#ifdef` variable for every single probe, and then probe-table.c will not be readable at all, which is usually rejected by people. In the original system of separate binaries this was solved easily dynamically by listing the contents of directory where the binaries could be found,
but in this prototype this seems to needed done at compilation time.

2. Second problem is more serious. The prototype doesn't work when it evaluates an XCCDF that references more than 1 OVAL file. For example https://github.com/OpenSCAP/openscap/blob/maint-1.2/tests/API/XCCDF/unittests/test_deriving_xccdf_result_from_oval.xccdf.xml references 2 OVALs.
For each OVAL a new OVAL session is created by the XCCDF session, which means multiple OVAL session exists at the same time, and for each session it creates a new set of probes.
If both OVALs contain the same type of OVAL objects  -- eg. both contain a file_object -- it runs a `file probe` 2 times in parellel. When the probes were separate processes, that wasn't a problem, because each process had its own adress space. But now, when probes are just threads, they share their global variables, which wasn't expected in the original probe design. When the both OVAL sessions are freed, the first free cleans the global variables of the probe, and the second free
causes a crash, because they are already freed. Also I think there can be other weird problems caused by sharing global variables between 2 probes of same type. The `file_probe` is a nice example of this behavior, but I believe we can find more examples in more other probes. This looks to me as unsolvable without either changing the XCCDF session logic or changing every probe to not use globals.

@mpreisler and others, what do you think can be a good solution for these 2 problems?

mpreisler commented:
> @mpreisler and others, what do you think can be a good solution for these 2 problems?

For the first one you already suggested a workable solution. As to the second problem, I am interested to know which of the 2 proposed solution do you prefer and why.

jan-cerny commented:
@mpreisler Thank you.     

> For the first one you already suggested a workable solution. 

Do you think that involving CMake and using C preprocessor is workable? How could the variables be produced and stored? What do you think about having something like `probe-table.h.in` from which CMake would generate `probe-table.h`? What do you suggest to do to make that readable and understandable for contributors?

> As to the second problem, I am interested to know which of the 2 proposed solution do you prefer and why.
Personally I think we should make sure the probes are reentrant.

Personally I don't prefer changing XCCDF session, because this PR already changes a lot in OVAL and I don't want to do changes in other modules if not necessary.

I would prefer that probes don't have global variables. The global variables are even forbidden for beginners.

Actually we could wrap the globals into 1 structure and probe can return a pointer to this structure in `probe_init`, this pointer is already part fo the probe context, which should be unique for each probe, and is already passed to probe_fini. I would like to try to figure out if this can work on the example of file probe.

jan-cerny commented on src/OVAL/oval_enumerations.c:
```diff
@@ -761,6 +779,17 @@ const char *oval_subtype_to_str(oval_subtype_t subtype)
 }
 
+oval_subtype_t oval_subtype_from_str(const char *str)
+{
```
This function cannot work and there was a good reason why there was no such function :smile:: One string can map to multiple OVAL subtypes. For example `"interface"` can map to `OVAL_IOS_INTERFACE` or `OVAL_UNIX_INTERFACE` or `OVAL_WINDOWS_INTERFACE`.

jan-cerny commented on src/OVAL/probes/SEAP/public/seap.h:
```diff
@@ -64,6 +65,7 @@ OSCAP_API int     SEAP_close (SEAP_CTX_t *ctx, int sd);
 
 OSCAP_API int SEAP_openfd (SEAP_CTX_t *ctx, int fd, uint32_t flags);
 OSCAP_API int SEAP_openfd2 (SEAP_CTX_t *ctx, int ifd, int ofd, uint32_t flags);
+OSCAP_API int SEAP_add_probe (SEAP_CTX_t *ctx, sch_queuedata_t *data);
```
code style

mpreisler commented on CMakeLists.txt:
```diff
@@ -163,6 +163,72 @@ option(ENABLE_PROBES_LINUX "build OVAL probes for the Linux OVAL tests" ${IS_LIN
+			endif()
+		endif()
+	endif()
+endif()
```
this approach won't work, you need to use dependent option as we discussed at the meeting. These can't be changed at configure time and the approach with the conditionals just can't work, it doesn't override to false, it overrides to true.

jan-cerny replied:
These options are already dependent on found dependencies (SELinux, DBus, RPM, ...). How it won't work? What do you suggest instead?

mpreisler replied:
They are not options, in other words you can't choose to disable any of the probes. This might be fine for this PR but we should do this properly in the future.

mpreisler commented on config.h.in:
```diff
@@ -84,6 +84,46 @@
 #cmakedefine HAVE_FLOCK
 #cmakedefine HAVE_STRPTIME
 
+#cmakedefine INDEPENDENT_ENVIRONMENTVARIABLE_PROBE
```
OPENSCAP_PROBE_*** would be a better scheme IMO, less clashing, prefix is better than suffix IMO

jan-cerny replied:
OK, I'll change that.

mpreisler commented on src/OVAL/probes/probe-table.c:
```diff
@@ -77,45 +193,123 @@ typedef struct probe_table_entry {
+#ifdef UNIX_XINETD_PROBE
 	{OVAL_UNIX_XINETD, xinetd_probe_init, xinetd_probe_main, xinetd_probe_fini, NULL},
-	{OVAL_INDEPENDENT_SYSCHAR_SUBTYPE, system_info_probe_init, system_info_probe_main, NULL, NULL},
+#endif
```
having to deal with each probe twice in this file is a pain, can't this be done more elegantly?

jan-cerny replied:
The problem is that to add a function name to this table we need to include a header file in which the function is defined. I thought about having a new header file in which we  include other header files, but then you would have to edit 2 files instead of 1 twice. Which is better?

mpreisler replied:
ok nvm. There probably is a solution where you can do both under one #ifdef but I don't know it off the top of my head.

jan-cerny replied:
Maybe we could have a macro that expands to `{OVAL_*SUBTYPE, *probe_init, *probe_main ...}` if something is defined and to an empty string if ti isn't defined, and the use it in this structure.

jan-cerny commented on src/OVAL/probes/SEAP/sch_queue.c:
```diff
@@ -0,0 +1,147 @@
+	pthread_mutex_t *mutex;
+	pthread_cond_t *cond;
+	int *cnt;
+	if (pthread_self() == data->parent_thread_id) {
```
This is wrong. We can't use `==` to compare thread ids, because on Windows, `pthread_t` is not a number but a structure. We need to use `pthread_equal()` instead.

jan-cerny commented on src/OVAL/probes/probe/probe_main.c:
```diff
@@ -184,57 +187,68 @@ static void preload_libraries_before_chroot()
-	oscap_set_verbose(verbosity_level, verbose_log_file, true);
 
-#ifdef _WIN32
+	pthread_setname_np(pthread_self(), "common_main");
```
missing `#ifdef HAVE_PTHREAD_SETNAME_NP`

jan-cerny commented on src/CMakeLists.txt:
```diff
@@ -86,11 +106,41 @@ if (BZIP2_FOUND)
+endif()
+if(SELINUX_FOUND)
+	target_link_libraries(openscap ${SELINUX_LIBRARIES})
+endif()
```
SELINUX_FOUND is twice here :cry:

jan-cerny commented:
This solution is terribly slow when evaluating some XCCDF rules from SCAP Security Guide (common Fedora profile). For example rule xccdf_org.ssgproject.content_rule_rpm_verify_permissions takes unbearable time. 

```
$ time oscap xccdf eval --rule xccdf_org.ssgproject.content
_rule_rpm_verify_permissions --profile common --results-arf /tmp/arf.xml /usr/share/xml/scap/ssg/content/
ssg-fedora-ds.xml                                                                                        
Title   Verify and Correct File Permissions with RPM                                                     
Rule    xccdf_org.ssgproject.content_rule_rpm_verify_permissions                                         
Result  fail              


real    6m31.640s         
user    6m20.823s         
sys     0m5.845s          
```

I wanted to profile that, but I couldn't get sprof working, so I tried OProfile http://oprofile.sourceforge.net which gives me this results for the command mentioned above:

```
samples  %        image name               app name                 symbol name
2476712  29.6118  libc-2.26.so             oscap                    getenv
1537886  18.3871  libopenscap.so.8.13.0    oscap                    __SEXP_VALIDATE
1089987  13.0320  libopenscap.so.8.13.0    oscap                    SEXP_val_dsc
828612    9.9070  libopenscap.so.8.13.0    oscap                    SEXP_rawval_lblk_cb
525330    6.2809  libc-2.26.so             oscap                    __strlen_avx2
328200    3.9240  libopenscap.so.8.13.0    oscap                    __SEXP_VALIDATE_cb
300243    3.5897  libopenscap.so.8.13.0    oscap                    SEXP_rawptr_bit
240442    2.8747  libopenscap.so.8.13.0    oscap                    SEXP_softrefp
```
I don't believe the results, because I think it spends time on mutexes, and I also haven't introduced any new call to getenv.

mpreisler commented:
@jan-cerny what numbers do you get with `maint-1.2` on the same machine?

jan-cerny commented:
@mpreisler something about 20 seconds

mpreisler commented:
> @mpreisler something about 20 seconds

That's quite a performance regression :-D Do you have any theories? Does the same thing happen with just one OVAL definition, just one XCCDF rule? In that case I would think there would be just one mutex locked and no significant races.

jan-cerny commented:
@mpreisler 
I expected it should be faster than before, because I have removed a lot of overhead - process forking, communication, ... 

This happens also if I just evaluate only the OVAL definition, eg. 
```
oscap oval eval --id oval:ssg-rpm_verify_permissions:def:1 /usr/share/xml/scap/ssg/content/ssg-fedora-oval.xml
```

This performance regression also happens on other resource intensive rules eg. xccdf_org.ssgproject.content_rule_rpm_verify_hashes or xccdf_org.ssgproject.content_rule_file_permissions_library_dirs

I don't have any theories why it become so slow. I wanted to find the reason, but it seems to me profiling doesn't help with this kind of problems. Do you have any suggestions? Could gdb somehow help? Or something else?

I think that the `getenv` in `__SEXP_validate` should be solved somehow else, because now it is called at almost every SEXP operation, it should be enough to call it only once and then remember the value in some bool variable. On the other hand I don't think this is the main problem, this has been there there before.

jan-cerny commented:
@mpreisler 

I have found that in fact there is no performance regression.

The `getenv` function of the `SEXP_validate` was the real reason of this slow down. The `getenv` is called in `SEXP_validate` which is called in most SEXP operations thousands times.
But SEXP_validate does something only if NDEBUG is not defined, otherwise SEXP_validate is defined as empty string. 

See https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/probes/SEAP/public/sexp-manip.h#L576

My mistake was that I compiled master with debugging symbols, but maint-1.2 without debugging symbols, so I compared incomparable things.

I think that we should somehow change this behavior to not slow down the code too much by SEXP validation. In debug mode it is run often, even in SEXP_free. But that change can be done in a separate PR.

jan-cerny commented:
Yes, with `CMAKE_BUILD_TYPE=Release` the performance is the same as it used to be:

```
time oscap xccdf eval --rule xccdf_org.ssgproject.content_rule_rpm_verify_permissions --profile common --results-arf /tmp/arf.xml /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml 
Title   Verify and Correct File Permissions with RPM
Rule    xccdf_org.ssgproject.content_rule_rpm_verify_permissions
Result  fail


real    0m22.251s
user    0m18.881s
sys     0m3.319s
```

What should we do with `SEXP_VALIDATE` in Debug mode?

jan-cerny commented:
My opinion is that we shouldn't check environment variables in every `SEXP_VALIDATE()` call and we shouldn't depend on CMake build type.

Instead, I would introduce a new CMake option, eg. `FULL_SEXP_VALIDATION`. This option will be turned off by default. Users will be able to turn it on at the configure time if they wish to debug SEXPs. This option would propagate to config.h and checked it the `SEXP_VALIDATE()` definition. If the option will be off, the SEXP_VAILDATE will do no operation.

mpreisler commented:
> My opinion is that we shouldn't check environment variables in every SEXP_VALIDATE() call and we shouldn't depend on CMake build type.
> 
> Instead, I would introduce a new CMake option, eg. FULL_SEXP_VALIDATION. This option will be turned off by default. Users will be able to turn it on at the configure time if they wish to debug SEXPs. This option would propagate to config.h and checked it the SEXP_VALIDATE() definition. If the option will be off, the SEXP_VALIDATE() will do no operation.

If it's this expensive I agree that this would make sense, similar to FULL_VALIDATION. Let's not do it in this PR though, let's not bloat the scope of this PR.

How close is this to being mergeable?

jan-cerny commented:
OK, I'll create a ticket for that.

It is comparable time, but I expected the new to be faster.

I compiled without debug symbols, and run common Fedora profile.
```
$ time oscap xccdf eval \
--profile common --results /tmp/results.xml \
--oval-results --report /tmp/report.html  \
 /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml
```
New:

```
real    3m2.793s
user    2m35.426s
sys     0m26.285s
```

Old from maint-1.2:

```
real    2m58.890s
user    2m30.932s
sys     0m27.231s
```

For merging this:
* Offline mode doesn't work. It doesn't work even for probes that don't use chroot, eg. rpmverifyfile probe.
* It crashes on Windows which I need to investigate.
* I haven't added Solaris probes.

jan-cerny commented:
The crashes on Windows that I experienced were caused by a wrong change in my local working copy. 

The problem is that there is still 1 occurrence of `posix_memalign`. Actually 2 occurences, one of them in used in normal mode and second one in debug mode. They can be found in `sm_memalign` or in `__sm_memalign_dbg` in file `sm_alloc.c`. See https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/probes/SEAP/sm_alloc.c

Windows doesn't have `posix_memalign`, so in my local copy I changed that to `malloc`, which was wrong and caused crashes. Some of the low level operation in SEXP module work with pointers and they depend on the pointer binary values, and assume the memory is aligned.
For example some of the functions in `sexp-value.c` See https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/probes/SEAP/sexp-value.c#L111

I think we can't get rid of these operations easily, because they are used widely in SEXP modules.

On Windows, MSVC 2017 has a function that can allocate aligned memory: `_aligned_malloc`. It has a problem that every pointer allocated by `_aligned_malloc` has to be freed by `_aligned_free`. If it is freed by normal `free` it crashes. See https://docs.microsoft.com/en-US/cpp/c-runtime-library/reference/aligned-free

In our SEXP implementation we don't distinguish between memory allocated by `posix_memalign` and `malloc`, and we free everything with `free`. It is very hard to track which memory was allocated aligned and which was normal, because SEXPs are freed at many places across the codebase.

I have done a small experiment. In Visual Studio I have replaced `posix_memalign` by `_aligned_malloc` in https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/probes/SEAP/sm_alloc.c#L203 and commented out `free` at L223 in the same file. With that it builds and doesn't crash. I have tried that by evaluating a simple OVAL with 1 family_object. But it is not a solution because it leaks memory.

mpreisler commented on compat/compat.h:
```diff
@@ -63,10 +63,10 @@ typedef SSIZE_T ssize_t;
+OSCAP_API extern char *optarg;
+OSCAP_API extern int optind;
+OSCAP_API extern int opterr;
+OSCAP_API extern int optopt;
```
so because of the name of the symbols nobody using getopt can use libopenscap at the same time?

jan-cerny replied:
I hope they can't, but I haven't verified.

jan-cerny commented:
@mpreisler I think that an easy solution will be to replace all malloc, frees and reallocs in file `sm-alloc.c` by their "aligned" counterparts. What do you think about that?

jan-cerny commented:
I tried my proposal and it doesn't work, because we don't use `sm_alloc` and `sm_free` consistently. Sometimes we allocate memory by standard `malloc` and free it by `sm_free`. See https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/probes/SEAP/sexp-manip_r.c#L261
Sometimes we do it vice versa, and sometimes the poiter at one place cal be allocated by both ways See https://github.com/OpenSCAP/openscap/blob/master/src/OVAL/oval_sexp.c#L928

What solution is good here?

jan-cerny commented:
What about having a table where we would store pointers to the memory allocated by `_aligned_malloc()` and when we will want to free something we would always check if the pointer is in that table, and if it's there we would use `_aligned_free()` otherwise `free()`?

mpreisler commented:
> What about having a table where we would store pointers to the memory allocated by _aligned_malloc() and when we will want to free something we would always check if the pointer is in that table, and if it's there we would use _aligned_free() otherwise free()?

That will work but there is considerable cost to doing this. IMO avoid this and do it cleanly, this is a crutch.

jan-cerny commented:
@openscap-ci test this please

mpreisler commented:
It passed! \o/

mpreisler commented:
It's odd that the performance is pretty much the same or a little worse than master. I would expect to get performance gains because we avoid IPC.

jan-cerny commented:
I wanted to fix the offline mode and modify the probes so that they don't use `chroot()` at all.
However I have found https://github.com/OpenSCAP/openscap/issues/1001 and then various problems described in https://github.com/OpenSCAP/openscap/pull/1002

My opinion is that implementing offline mode without `chroot()` correctly requires fundamental changes across whole OVAL logic, namely in SEXP module. I think that is a big task which is out of scope of the Windows effort. @mpreisler What do you think?

jan-cerny commented:
Blocked by https://github.com/OpenSCAP/openscap/pull/1000 and merging maint-1.2 to master.

mpreisler commented:
@openscap-ci test this please

mpreisler commented:
ACK

Great work!
